pkg/controller/srcdst_controller.go

```diff
@@ -4,110 +4,154 @@ import (
+// check on EC2 instances.
+func NewSrcDstController(client kubernetes.Interface,
+	nodeInformer informer.NodeInformer,
+	ec2Client *ec2.EC2) *Controller {
+
 	c := &Controller{
-		client:    client,
-		ec2Client: ec2Client,
+		client:       client,
+		nodeInformer: nodeInformer.Informer(),
+		nodeSynced:   nodeInformer.Informer().HasSynced,
+		workqueue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Nodes"),
+		ec2Client:    ec2Client,
 	}
 
-	nodeListWatcher := cache.NewListWatchFromClient(
-		client.Core().RESTClient(),
-		"nodes",
-		v1.NamespaceAll,
-		fields.Everything())
-
-	_, controller := cache.NewInformer(
-		nodeListWatcher,
-		&v1.Node{},
-		60*time.Second,
-		// Callback Functions to trigger on add/update/delete
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    c.handler,
-			UpdateFunc: func(old, new interface{}) { c.handler(new) },
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.handler,
+		UpdateFunc: func(old, new interface{}) {
+			c.handler(new)
 		},
-	)
-
-	c.Controller = controller
+	})
 
 	return c
 }
 
+func (c *Controller) Run(numWorkers int, stopCh <-chan struct{}) error {
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	if ok := cache.WaitForCacheSync(stopCh, c.nodeSynced); !ok {
+		return fmt.Errorf("caches have failed to sync")
+	}
+
+	var wg sync.WaitGroup
+	for i := 0; i < numWorkers; i++ {
+		c.wg.Add(1)
+		go func() {
+			defer wg.Done()
+			go wait.Until(c.runWorker, time.Second, stopCh)
+		}()
+	}
+
+	c.wg.Wait()
+	<-stopCh
+	return nil
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.workqueue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.workqueue.Forget(key)
+	if srcDstEnabled, err := c.checkSrcDstAttributeEnabled(key.(string)); err != nil {
+		klog.Error(err)
+		c.workqueue.AddRateLimited(key)
+	} else if srcDstEnabled {
+		return true
+	}
+
+	if err := c.disableSrcDstCheck(key.(string)); err != nil {
+		c.workqueue.AddRateLimited(key)
+	}
+
+	return true
+}
+
 func (c *Controller) handler(obj interface{}) {
-	// this handler makes sure that all nodes within a cluster has its src/dst check disabled in EC2
-	node, ok := obj.(*v1.Node)
-	if !ok {
-		glog.Errorf("Expected Node but handler received: %+v", obj)
+
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
 		return
 	}
-	glog.V(4).Infof("Received update of node: %s", node.Name)
-	c.disableSrcDstIfEnabled(node)
+	c.workqueue.Add(key)
+	return
 }
 
-func (c *Controller) disableSrcDstIfEnabled(node *v1.Node) {
-	srcDstCheckEnabled := true
-	if node.Annotations != nil {
-		if _, ok := node.Annotations[SrcDstCheckDisabledAnnotation]; ok {
-			srcDstCheckEnabled = false
+func (c *Controller) disableSrcDstCheck(key string) error {
+	defer c.workqueue.Done(key)
+
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
```

Comment:
is it possible to have a less greedy retry that does not calling 'c.modifySrcDstCheckAttribute' again if there's a conflict?